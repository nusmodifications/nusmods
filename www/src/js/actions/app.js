// @flow

import type { FSA } from 'types/redux';
import type { NotificationOptions } from 'types/reducers';

export const SET_ONLINE_STATUS = 'SET_ONLINE_STATUS';
export function setOnlineStatus(isOnline: boolean): FSA {
  return {
    type: SET_ONLINE_STATUS,
    payload: { isOnline },
  };
}

export const TOGGLE_FEEDBACK_MODAL = 'TOGGLE_FEEDBACK_MODAL';
export function toggleFeedback(): FSA {
  return {
    type: TOGGLE_FEEDBACK_MODAL,
    payload: null,
  };
}

export const OPEN_NOTIFICATION = 'OPEN_NOTIFICATION';
export function openNotification(message: string, options: NotificationOptions = {}): FSA {
  return {
    type: OPEN_NOTIFICATION,
    payload: {
      message,
      ...options,
    },
  };
}

export const POP_NOTIFICATION = 'POP_NOTIFICATION';
export function popNotification(): FSA {
  return {
    type: POP_NOTIFICATION,
    payload: null,
  };
}
