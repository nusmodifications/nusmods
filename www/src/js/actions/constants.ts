/** moduleBank constants * */

export const FETCH_MODULE = 'FETCH_MODULE'; // Action to fetch modules
export const FETCH_MODULE_LIST = 'FETCH_MODULE_LIST';
export const UPDATE_MODULE_TIMESTAMP = 'UPDATE_MODULE_TIMESTAMP';
export const REMOVE_LRU_MODULE = 'REMOVE_LRU_MODULE';
export const FETCH_ARCHIVE_MODULE = 'FETCH_ARCHIVE_MODULE'; // Action to fetch module from previous years

/** undoHistory constants * */

export const UNDO = 'UNDO';
export const REDO = 'REDO';

/** export constant(s) * */

export const SET_EXPORTED_DATA = 'SET_EXPORTED_DATA';
