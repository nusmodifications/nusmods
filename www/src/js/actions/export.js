// @flow
import type { Module, Semester } from 'types/modules';

import { hydrateSemTimetableWithLessons } from 'utils/timetables';
import type { ExportData } from 'types/export';
import type { FSA, GetState } from 'types/redux';
import { getSemesterTimetable } from 'reducers/timetables';

function downloadUrl(blob: Blob, filename: string) {
  const link = document.createElement('a');
  if (window.navigator.msSaveOrOpenBlob) {
    window.navigator.msSaveOrOpenBlob(blob, filename);
  } else {
    const objectUrl = URL.createObjectURL(blob);
    link.download = filename;
    link.href = objectUrl;
    link.dispatchEvent(new MouseEvent('click'));
    URL.revokeObjectURL(objectUrl);
  }
}

export const SUPPORTS_DOWNLOAD = 'download' in document.createElement('a');

export function downloadAsIcal(semester: Semester) {
  return (dispatch: Function, getState: GetState) => {
    Promise.all([
      import(/* webpackChunkName: "export" */ 'ical-generator'),
      import(/* webpackChunkName: "export" */ 'utils/ical'),
    ]).then(([ical, icalUtils]) => {
      const { moduleBank, timetables } = getState();
      const modules = moduleBank.modules;
      const { timetable } = getSemesterTimetable(semester, timetables);
      const timetableWithLessons = hydrateSemTimetableWithLessons(timetable, modules, semester);

      const events = icalUtils.default(semester, timetableWithLessons, modules);
      // $FlowFixMe Flow doesn't seem to like import() for CJS modules
      const cal = ical({
        domain: 'nusmods.com',
        prodId: '//NUSMods//NUSMods//EN',
        events,
      });

      const blob = new Blob([cal.toString()], { type: 'text/plain' });
      downloadUrl(blob, 'nusmods_calendar.ics');
    });
  };
}

export const SET_EXPORTED_DATA = 'SET_EXPORTED_DATA';
export function setExportedData(modules: Module[], data: ExportData): FSA {
  return {
    type: SET_EXPORTED_DATA,
    payload: {
      modules,
      ...data,
    },
  };
}
