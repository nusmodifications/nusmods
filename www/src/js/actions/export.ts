// @flow

import type { Module, Semester } from 'types/modules';
import type { ExportData } from 'types/export';
import type { FSA, GetState } from 'types/redux';
import { hydrateSemTimetableWithLessons } from 'utils/timetables';
import { captureException, retryImport } from 'utils/error';
import { getSemesterTimetable } from 'reducers/timetables';

function downloadUrl(blob: Blob, filename: string) {
  const link = document.createElement('a');
  if (window.navigator.msSaveOrOpenBlob) {
    window.navigator.msSaveOrOpenBlob(blob, filename);
  } else {
    const objectUrl = URL.createObjectURL(blob);
    link.download = filename;
    link.href = objectUrl;
    link.dispatchEvent(new MouseEvent('click'));
    URL.revokeObjectURL(objectUrl);
  }
}

export const SUPPORTS_DOWNLOAD = 'download' in document.createElement('a');

export function downloadAsIcal(semester: Semester) {
  return (dispatch: Function, getState: GetState) => {
    Promise.all([
      retryImport(() => import(/* webpackChunkName: "export" */ 'ical-generator')),
      retryImport(() => import(/* webpackChunkName: "export" */ 'utils/ical')),
    ])
      .then(([ical, icalUtils]) => {
        const {
          moduleBank: { modules },
          timetables,
        } = getState();
        const { timetable } = getSemesterTimetable(semester, timetables);
        const timetableWithLessons = hydrateSemTimetableWithLessons(timetable, modules, semester);

        const events = icalUtils.default(semester, timetableWithLessons, modules);
        const cal = ical.default({
          domain: 'nusmods.com',
          prodId: '//NUSMods//NUSMods//EN',
          events,
        });

        const blob = new Blob([cal.toString()], { type: 'text/plain' });
        downloadUrl(blob, 'nusmods_calendar.ics');
      })
      .catch(captureException);
  };
}

export const SET_EXPORTED_DATA = 'SET_EXPORTED_DATA';
export function setExportedData(modules: Module[], data: ExportData): FSA {
  return {
    type: SET_EXPORTED_DATA,
    payload: {
      modules,
      ...data,
    },
  };
}
