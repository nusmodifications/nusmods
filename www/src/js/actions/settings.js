// @flow
import type { FSA } from 'types/redux';
import type { Faculty, Semester } from 'types/modules';
import type { Mode } from 'types/settings';

export const SELECT_SEMESTER: string = 'SELECT_SEMESTER';
export function selectSemester(semester: Semester): FSA {
  return {
    type: SELECT_SEMESTER,
    payload: semester,
  };
}

export const SELECT_NEW_STUDENT: string = 'SELECT_NEW_STUDENT';
export function selectNewStudent(newStudent: boolean): FSA {
  return {
    type: SELECT_NEW_STUDENT,
    payload: newStudent,
  };
}

export const SELECT_FACULTY: string = 'SELECT_FACULTY';
export function selectFaculty(faculty: Faculty): FSA {
  return {
    type: SELECT_FACULTY,
    payload: faculty,
  };
}

export const SELECT_MODE: string = 'SELECT_MODE';
export function selectMode(mode: Mode): FSA {
  return {
    type: SELECT_MODE,
    payload: mode,
  };
}

export const TOGGLE_MODE: string = 'TOGGLE_MODE';
export function toggleMode(): FSA {
  return {
    type: TOGGLE_MODE,
    payload: null,
  };
}

export const DISMISS_CORS_NOTIFICATION = 'DISMISS_CORS_NOTIFICATION';
export function dismissCorsNotification(round: string): FSA {
  return {
    type: DISMISS_CORS_NOTIFICATION,
    payload: { round },
  };
}

export const ENABLE_CORS_NOTIFICATION = 'ENABLE_CORS_NOTIFICATION';
export function enableCorsNotification(round: string): FSA {
  return {
    type: ENABLE_CORS_NOTIFICATION,
    payload: { round },
  };
}

export const TOGGLE_CORS_NOTIFICATION_GLOBALLY = 'TOGGLE_CORS_NOTIFICATION_GLOBALLY';
export function toggleCorsNotificationGlobally(enabled: boolean): FSA {
  return {
    type: TOGGLE_CORS_NOTIFICATION_GLOBALLY,
    payload: { enabled },
  };
}
