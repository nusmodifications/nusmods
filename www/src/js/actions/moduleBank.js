// @flow
import type { AcadYear, Module, ModuleCode } from 'types/modules';
import type { TimetableConfig } from 'types/timetables';

import type { GetState } from 'types/redux';
import type { ModulesMap } from 'reducers/moduleBank';
import { size, get, flatMap, sortBy, zip } from 'lodash';
import { requestAction } from 'actions/requests';
import NUSModsApi from 'apis/nusmods';
import config from 'config';

const MAX_MODULE_LIMIT: number = 100;

export const FETCH_MODULE_LIST: string = 'FETCH_MODULE_LIST';
export function fetchModuleList() {
  return requestAction(FETCH_MODULE_LIST, FETCH_MODULE_LIST, {
    url: NUSModsApi.moduleListUrl(),
  });
}

// Action to fetch modules
export const FETCH_MODULE: string = 'FETCH_MODULE';
export function fetchModuleRequest(moduleCode: ModuleCode) {
  return `${FETCH_MODULE}/${moduleCode}`;
}

export function getRequestModuleCode(key: string): ?ModuleCode {
  const parts = key.split('/');
  if (parts.length === 2 && parts[0] === FETCH_MODULE) return parts[1];
  return null;
}

export const UPDATE_MODULE_TIMESTAMP = 'UPDATE_MODULE_TIMESTAMP';
export function updateModuleTimestamp(moduleCode: ModuleCode) {
  return {
    type: UPDATE_MODULE_TIMESTAMP,
    payload: moduleCode,
  };
}

export const REMOVE_LRU_MODULE = 'REMOVE_LRU_MODULE';
export function removeLRUModule(moduleCodes: ModuleCode[]) {
  return {
    type: REMOVE_LRU_MODULE,
    payload: moduleCodes,
  };
}

// Export for testing
export function getLRUModules(
  modules: ModulesMap,
  lessons: TimetableConfig,
  currentModule: string,
  toRemove: number = 1,
): ModuleCode[] {
  // Pull all the modules in all the timetables
  const timetableModules = new Set(flatMap(lessons, (semester) => Object.keys(semester)));

  // Remove the module which is least recently used and which is not in timetable
  // and not the currently loaded one
  const canRemove = Object.keys(modules).filter(
    (moduleCode) => moduleCode !== currentModule && !timetableModules.has(moduleCode),
  );

  // Sort them based on the timestamp alone
  const sortedModules = sortBy(canRemove, (moduleCode) =>
    get(modules[moduleCode], ['timestamp'], 0),
  );

  return sortedModules.slice(0, toRemove);
}

export function fetchModule(moduleCode: ModuleCode) {
  return (dispatch: Function, getState: GetState) => {
    const onFinally = () => {
      // Update the timestamp of the accessed module if it is in the store.
      if (getState().moduleBank.modules[moduleCode]) {
        dispatch(updateModuleTimestamp(moduleCode));
      }

      // Remove the LRU module if the size exceeds the maximum and if anything
      // can be removed
      const overLimitCount = size(getState().moduleBank.modules) - MAX_MODULE_LIMIT;
      if (overLimitCount > 0) {
        const { moduleBank, timetables } = getState();

        const LRUModule = getLRUModules(
          moduleBank.modules,
          timetables.lessons,
          moduleCode,
          overLimitCount,
        );

        if (LRUModule) {
          dispatch(removeLRUModule(LRUModule));
        }
      }
    };

    const key = fetchModuleRequest(moduleCode);
    return dispatch(
      requestAction(key, FETCH_MODULE, {
        url: NUSModsApi.moduleDetailsUrl(moduleCode),
      }),
    ).then(
      (result) => {
        onFinally();
        return result;
      },
      (error) => {
        onFinally();
        throw error;
      },
    );
  };
}

// Action to fetch module from previous years
export const FETCH_ARCHIVE_MODULE: string = 'FETCH_ARCHIVE_MODULE';
export function fetchArchiveRequest(moduleCode: ModuleCode, year: string) {
  return `${FETCH_ARCHIVE_MODULE}_${moduleCode}_${year}`;
}

export function fetchModuleArchive(moduleCode: ModuleCode, year: string) {
  const key = fetchArchiveRequest(moduleCode, year);
  const action = requestAction(key, FETCH_ARCHIVE_MODULE, {
    url: NUSModsApi.moduleDetailsUrl(moduleCode, year),
  });

  action.meta.academicYear = year;

  return action;
}

export function fetchAllModuleArchive(moduleCode: ModuleCode) {
  // Returns: Promise<[AcadYear, Module?][]>
  return (dispatch: Function) =>
    Promise.all(
      config.archiveYears.map((year) =>
        dispatch(fetchModuleArchive(moduleCode, year)).catch(() => null),
      ),
    ).then((modules) => zip<AcadYear, Array<Module>>(config.archiveYears, modules));
}
