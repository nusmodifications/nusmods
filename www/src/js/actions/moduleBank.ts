import { requestAction } from 'actions/requests';
import NUSModsApi from 'apis/nusmods';
import config from 'config';
import { flatMap, get, size, sortBy, zip } from 'lodash';
import { ModulesMap } from 'types/reducers';
import { AcadYear, Module, ModuleCode } from 'types/modules';

import { GetState } from 'types/redux';
import { TimetableConfig } from 'types/timetables';
import {
  FETCH_ARCHIVE_MODULE,
  FETCH_MODULE,
  FETCH_MODULE_LIST,
  REMOVE_LRU_MODULE,
  UPDATE_MODULE_TIMESTAMP,
} from './constants';

const MAX_MODULE_LIMIT = 100;

export function fetchModuleList() {
  return requestAction(FETCH_MODULE_LIST, FETCH_MODULE_LIST, {
    url: NUSModsApi.moduleListUrl(),
  });
}

export function fetchModuleRequest(moduleCode: ModuleCode) {
  return `${FETCH_MODULE}/${moduleCode}`;
}

export function getRequestModuleCode(key: string): ModuleCode | null {
  const parts = key.split('/');
  if (parts.length === 2 && parts[0] === FETCH_MODULE) return parts[1];
  return null;
}

export function updateModuleTimestamp(moduleCode: ModuleCode) {
  return {
    type: UPDATE_MODULE_TIMESTAMP,
    payload: moduleCode,
  };
}

export function removeLRUModule(moduleCodes: ModuleCode[]) {
  return {
    type: REMOVE_LRU_MODULE,
    payload: moduleCodes,
  };
}

// Export for testing
export function getLRUModules(
  modules: ModulesMap,
  lessons: TimetableConfig,
  currentModule: string,
  toRemove: number = 1,
): ModuleCode[] {
  // Pull all the modules in all the timetables
  const timetableModules = new Set(flatMap(lessons, (semester) => Object.keys(semester)));

  // Remove the module which is least recently used and which is not in timetable
  // and not the currently loaded one
  const canRemove: ModuleCode[] = Object.keys(modules).filter(
    (moduleCode) => moduleCode !== currentModule && !timetableModules.has(moduleCode),
  );

  // Sort them based on the timestamp alone
  const sortedModules = sortBy<ModuleCode>(canRemove, (moduleCode) =>
    get(modules[moduleCode], ['timestamp'], 0),
  );

  return sortedModules.slice(0, toRemove);
}

export function fetchModule(moduleCode: ModuleCode) {
  return (dispatch: Function, getState: GetState) => {
    const onFinally = () => {
      // Update the timestamp of the accessed module if it is in the store.
      if (getState().moduleBank.modules[moduleCode]) {
        dispatch(updateModuleTimestamp(moduleCode));
      }

      // Remove the LRU module if the size exceeds the maximum and if anything
      // can be removed
      const overLimitCount = size(getState().moduleBank.modules) - MAX_MODULE_LIMIT;
      if (overLimitCount > 0) {
        const { moduleBank, timetables } = getState();

        const LRUModule = getLRUModules(
          moduleBank.modules,
          timetables.lessons,
          moduleCode,
          overLimitCount,
        );

        if (LRUModule) {
          dispatch(removeLRUModule(LRUModule));
        }
      }
    };

    const key = fetchModuleRequest(moduleCode);
    return dispatch(
      requestAction(key, FETCH_MODULE, {
        url: NUSModsApi.moduleDetailsUrl(moduleCode),
      }),
    ).then(
      (result: any) => {
        onFinally();
        return result;
      },
      (error: any) => {
        onFinally();
        throw error;
      },
    );
  };
}

export function fetchArchiveRequest(moduleCode: ModuleCode, year: string) {
  return `${FETCH_ARCHIVE_MODULE}_${moduleCode}_${year}`;
}

export function fetchModuleArchive(moduleCode: ModuleCode, year: string) {
  const key = fetchArchiveRequest(moduleCode, year);
  const action = requestAction(key, FETCH_ARCHIVE_MODULE, {
    url: NUSModsApi.moduleDetailsUrl(moduleCode, year),
  });

  action.meta.academicYear = year;

  return action;
}

export function fetchAllModuleArchive(moduleCode: ModuleCode) {
  // Returns: Promise<[AcadYear, Module?][]>
  return (dispatch: Function) =>
    Promise.all(
      config.archiveYears.map((year) =>
        dispatch(fetchModuleArchive(moduleCode, year)).catch(() => null),
      ),
    ).then((modules) => zip<AcadYear, Module[]>(config.archiveYears, modules));
}
