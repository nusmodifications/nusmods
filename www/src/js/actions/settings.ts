import { FSA } from 'types/redux';
import { Faculty, Semester } from 'types/modules';
import { Mode } from 'types/settings';
import { ModuleTableOrder } from 'types/views';

export const SELECT_SEMESTER = 'SELECT_SEMESTER';
export function selectSemester(semester: Semester): FSA {
  return {
    type: SELECT_SEMESTER,
    payload: semester,
  };
}

export const SELECT_NEW_STUDENT = 'SELECT_NEW_STUDENT';
export function selectNewStudent(newStudent: boolean): FSA {
  return {
    type: SELECT_NEW_STUDENT,
    payload: newStudent,
  };
}

export const SELECT_FACULTY = 'SELECT_FACULTY';
export function selectFaculty(faculty: Faculty): FSA {
  return {
    type: SELECT_FACULTY,
    payload: faculty,
  };
}

export const SELECT_MODE = 'SELECT_MODE';
export function selectMode(mode: Mode): FSA {
  return {
    type: SELECT_MODE,
    payload: mode,
  };
}

export const TOGGLE_MODE = 'TOGGLE_MODE';
export function toggleMode(): FSA {
  return {
    type: TOGGLE_MODE,
    payload: null,
  };
}

export const DISMISS_CORS_NOTIFICATION = 'DISMISS_CORS_NOTIFICATION';
export function dismissCorsNotification(round: string): FSA {
  return {
    type: DISMISS_CORS_NOTIFICATION,
    payload: { round },
  };
}

export const ENABLE_CORS_NOTIFICATION = 'ENABLE_CORS_NOTIFICATION';
export function enableCorsNotification(round: string): FSA {
  return {
    type: ENABLE_CORS_NOTIFICATION,
    payload: { round },
  };
}

export const TOGGLE_CORS_NOTIFICATION_GLOBALLY = 'TOGGLE_CORS_NOTIFICATION_GLOBALLY';
export function toggleCorsNotificationGlobally(enabled: boolean): FSA {
  return {
    type: TOGGLE_CORS_NOTIFICATION_GLOBALLY,
    payload: { enabled },
  };
}

export const SET_MODULE_TABLE_SORT = 'SET_MODULE_TABLE_SORT';
export function setModuleTableOrder(sort: ModuleTableOrder): FSA {
  return {
    type: SET_MODULE_TABLE_SORT,
    payload: sort,
  };
}

export const TOGGLE_BETA_TESTING_STATUS = 'TOGGLE_BETA_TESTING_STATUS';
export function toggleBetaTesting(): FSA {
  return {
    type: TOGGLE_BETA_TESTING_STATUS,
    payload: null,
  };
}

export const SET_LOAD_DISQUS_MANUALLY = 'SET_DISQUS_MANUAL_LOAD';
export function setLoadDisqusManually(status: boolean): FSA {
  return {
    type: SET_LOAD_DISQUS_MANUALLY,
    payload: status,
  };
}
