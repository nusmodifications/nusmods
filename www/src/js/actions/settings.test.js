// @flow
import type { Faculty, Semester } from 'types/modules';

import * as actions from 'actions/settings';

import { LIGHT_MODE } from 'types/settings';

describe('settings', () => {
  test('should dispatch a select of a semester value', () => {
    const semester: Semester = 1;
    expect(actions.selectSemester(semester)).toMatchSnapshot();
  });

  test('should dispatch a select of a newStudent value', () => {
    const newStudent = false;
    expect(actions.selectNewStudent(newStudent)).toMatchSnapshot();
  });

  test('should dispatch a selection of a mode', () => {
    expect(actions.selectMode(LIGHT_MODE)).toMatchSnapshot();
  });

  test('should dispatch a toggle of a mode', () => {
    expect(actions.toggleMode()).toMatchSnapshot();
  });

  test('should dispatch a select of a faculty value', () => {
    const faculty: Faculty = 'Some faculty';
    expect(actions.selectFaculty(faculty)).toMatchSnapshot();
  });
});
