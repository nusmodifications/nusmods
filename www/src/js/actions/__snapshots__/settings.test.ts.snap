// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings should dispatch a select of a faculty value 1`] = `
Object {
  "payload": "Some faculty",
  "type": "SELECT_FACULTY",
}
`;

exports[`settings should dispatch a select of a newStudent value 1`] = `
Object {
  "payload": false,
  "type": "SELECT_NEW_STUDENT",
}
`;

exports[`settings should dispatch a select of a semester value 1`] = `
Object {
  "payload": 1,
  "type": "SELECT_SEMESTER",
}
`;

exports[`settings should dispatch a selection of a mode 1`] = `
Object {
  "payload": "LIGHT",
  "type": "SELECT_MODE",
}
`;

exports[`settings should dispatch a toggle of a mode 1`] = `
Object {
  "payload": null,
  "type": "TOGGLE_MODE",
}
`;
